### PR TITLE
initial attempt at increased tolerance for daemonsets

### PR DESCRIPTION
**Pull request by dalehamel**
**What are you trying to accomplish with this PR?**

Provide a solution to issue #565 

**How is this accomplished?**

    Support treating daemonsets as successful if new nodes added
    
    This allows for the daemonset to be treated as successful even if
    the daemonset is not 100% converged, so long as:
    
    - The desired and scheduled number of pods match
    - All desired pods for which there was an associated node at the
    start of the deploy process are in the Ready state
    
    This should allow for better toleration of autoscaling events, which
    can currently prevent convergence on daemonsets as new nodes may be
    added in the middle of the deploy.

**What could go wrong?**

If I am missing edge cases, it is possible that deploys could be prematurely and incorrectly marked as successful, which is not desirable.

### CODE REVIEW

**Comment by dalehamel** [comment]
There is more subtlety here than I realized - we can't rely on the assumption that every daemonset is scheduled on every node, so we need to consider this based on the pods instead rather than based on the nodes.

We'll take the list of nodes at the start of the deployment as the only ones under consideration. If we have pods that aren't ready, which are scheduled on nodes that aren't in this initial list, we can safely ignore them. 

The set difference of the nodes_for_pods - original_nodes_list should be a strict superset. There should be no nodes in the original list for which there is not a ready pod that is scheduled on that node.

File: lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb [hunk 1]
```diff
@@ -17,8 +18,15 @@ def status
 
     def deploy_succeeded?
       return false unless exists?
-      rollout_data["desiredNumberScheduled"].to_i == rollout_data["updatedNumberScheduled"].to_i &&
-      rollout_data["desiredNumberScheduled"].to_i == rollout_data["numberReady"].to_i &&
+
+      puts pods.inspect
+      puts "TOTAL_PODS #{@pods.size}"
+      considered_pods = @pods.select { |p| @nodes.map(&:name).include?(p.node_name) }
```
Comment (dalehamel) [inline lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb]: now that I think about this more, this check should probably be gated by `rollout_data["desiredNumberScheduled"].to_i == rollout_data["updatedNumberScheduled"].to_i`

File: .gitignore [hunk 2]
```diff
@@ -18,3 +18,5 @@ dev/flamegraph.svg
 dev/profile
 dev/flamegraph.pl
 .local-context
+*.swp
```
Comment (dalehamel) [inline .gitignore]: sorry, vim stuff :nerd_face:

File: lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb [hunk 3]
```diff
@@ -17,8 +18,11 @@ def status
 
     def deploy_succeeded?
       return false unless exists?
-      rollout_data["desiredNumberScheduled"].to_i == rollout_data["updatedNumberScheduled"].to_i &&
-      rollout_data["desiredNumberScheduled"].to_i == rollout_data["numberReady"].to_i &&
+
+      considered_pods = @pods.select { |p| @nodes.map(&:name).include?(p.node_name) }
+      (rollout_data["desiredNumberScheduled"].to_i == rollout_data["updatedNumberScheduled"].to_i &&
```
Comment (dalehamel) [inline lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb]: ok so i am well aware this multi-line complicated conditional isn't the prettiest thing in the world, but hey rubocop allowed it!

Open to ideas on how to make this clearer if the boolean logic is getting to sophisticated.

File: lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb [hunk 4]
```diff
@@ -38,6 +42,31 @@ def print_debug_logs?
...
+      end
+
+      def initialize(definition:)
+        @name = definition.dig("metadata", "name").to_s
+        @definition = definition
+      end
+    end
+
+    def find_nodes(cache)
+      # FIXME: should this also exclude tainted nodes?
```
Comment (dalehamel) [inline lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb]: NIT - remove this before merging, i don't think we care about tainted nodes

File: lib/kubernetes-deploy/kubernetes_resource/pod.rb [hunk 5]
```diff
@@ -101,6 +101,16 @@ def print_debug_logs?
       exists? && !@stream_logs # don't print them a second time
     end
 
+    def ready?
```
Comment (dalehamel) [inline lib/kubernetes-deploy/kubernetes_resource/pod.rb]: I had to pull this out of `private`, so I guess i'm the first one to have a valid reason to call it. I think this is fine as a public method.

File: test/fixtures/for_unit_tests/daemon_set.yml [hunk 6]
```diff
@@ -11,7 +11,7 @@ spec:
   template:
     metadata:
       labels:
-        app: hello-cloud
```
Comment (dalehamel) [inline test/fixtures/for_unit_tests/daemon_set.yml]: I think this was an existing bug

File: test/fixtures/for_unit_tests/node.yml [hunk 7]
```diff
@@ -0,0 +1,136 @@
+apiVersion: v1
```
Comment (dalehamel) [inline test/fixtures/for_unit_tests/node.yml]: straight out of minikube

File: test/fixtures/for_unit_tests/daemon_set_pod.yml [hunk 8]
```diff
@@ -0,0 +1,105 @@
+apiVersion: v1
```
Comment (dalehamel) [inline test/fixtures/for_unit_tests/daemon_set_pod.yml]: Result from minikube when I applied the DS yaml.

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 9]
```diff
@@ -4,9 +4,11 @@
 class DaemonSetTest < KubernetesDeploy::TestCase
   include ResourceCacheTestHelper
 
+  OWNER_UID = 'c31a9b4e-e6dd-11e9-8f47-e6322f98393a'
```
Comment (KnVerey) [inline test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: Could we instead make the UID match in the fixtures themselves?

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 10]
```diff
@@ -38,22 +40,78 @@ def test_deploy_passes_when_updated_available_does_match
...
+    else
+      [base_node_manifest].flatten
+    end
   end
 
-  def build_synced_ds(template:)
-    ds = KubernetesDeploy::DaemonSet.new(namespace: "test", context: "nope", logger: logger, definition: template)
-    stub_kind_get("DaemonSet", items: [template])
-    stub_kind_get("Pod", items: [])
+  def build_synced_ds(ds_template:, pod_templates: nil, node_templates: nil)
```
Comment (KnVerey) [inline test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: Suggestion: since the keywords are plural, make the default values `[]` and then remove the `|| []` below

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 11]
```diff
@@ -38,22 +40,78 @@ def test_deploy_passes_when_updated_available_does_match
...
+
+  def test_deploy_passes_when_ready_pods_for_multiple_nodes
+    status = {
+      "desiredNumberScheduled": 3,
+      "updatedNumberScheduled": 3,
+      "numberReady": 2,
+    }
+    ds_template = build_ds_template(filename: 'daemon_set.yml', uid: OWNER_UID, generation: 2, status: status)
+    pod_template = build_pod_template(filename: 'daemon_set_pod.yml')
+    node_template = build_node_template(filename: 'node.yml')
```
Comment (KnVerey) [inline test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: 3 desired, 2 ready and 1 node seems like a weird state that would never be observed on a single sync cycle--you'd for sure see three nodes given that status. If we go with the node list caching approach, this test will need to multiple sync cycles. Otherwise, it will need to use multiple nodes with manipulated creation timestamps.

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 12]
```diff
@@ -38,22 +40,78 @@ def test_deploy_passes_when_updated_available_does_match
...
 
+  def test_deploy_fails_when_not_all_pods_updated
+    status = {
+      "desiredNumberScheduled": 3,
+      "updatedNumberScheduled": 2,
+      "numberReady": 2,
+    }
+    ds_template = build_ds_template(filename: 'daemon_set.yml', uid: OWNER_UID, generation: 2, status: status)
+    pod_template = build_pod_template(filename: 'daemon_set_pod.yml')
+    node_template = build_node_template(filename: 'node.yml')
```
Comment (KnVerey) [inline test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: Same comment about this test.

File: lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb (hunk 4 above)
Comment (KnVerey) [reply lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb]: Agreed, we don't care about tainted nodes. If the pod was scheduled, we should assume it was supposed to be.

File: lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb [hunk 13]
```diff
@@ -8,6 +8,7 @@ class DaemonSet < PodSetBase
     def sync(cache)
       super
       @pods = exists? ? find_pods(cache) : []
+      @nodes = find_nodes(cache)
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb]: going with the first approach + `presence` as you suggested IRL

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb (hunk 11 above)
Comment (dalehamel) [reply test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: I agree these states are weird. the intent was to simulate where a node has been added by autoscaling. I'm going to adjust the numbers accordingly.

> If we go with the node list caching approach, this test will need to multiple sync cycles

I can add a separate test that handles the multiple-sync case

Comment (dalehamel) [reply test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: I think that `test_deploy_passes_when_ready_pods_but_node_added` should cover this, but i wasn't sure how I could use the same daemonset instance, though I'm not totally sure it matters.

Can you think of a way to reuse the daemonset instance and update the status transparently? Ultimately to test logic around cache I think that's needed.

**Comment by dalehamel** [comment]
thanks for the review @KnVerey I think i addressed everything, can you do another pass?

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 14]
```diff
@@ -38,22 +38,105 @@ def test_deploy_passes_when_updated_available_does_match
...
+
+    node_added_status = {
+      "desiredNumberScheduled": 3,
+      "updatedNumberScheduled": 3,
+      "numberReady": 1,
+    }
+    ds_template = build_ds_template(filename: 'daemon_set.yml', generation: 2, status: node_added_status)
+    pod_template = build_pod_template(filename: 'daemon_set_pod.yml')
+    node_template = build_node_template(filename: 'nodes.yml')
+    ds = build_synced_ds(ds_template: ds_template, pod_templates: pod_template, node_templates: node_template)
```
Comment (KnVerey) [inline test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: > I think that test_deploy_passes_when_ready_pods_but_node_added should cover this, but i wasn't sure how I could use the same daemonset instance, though I'm not totally sure it matters.

> Can you think of a way to reuse the daemonset instance and update the status transparently? Ultimately to test logic around cache I think that's needed.

This is the line that needs to change--you need to manually do this method's job on your old instance:

```ruby
    stub_kind_get("DaemonSet", items: [ds_template]) # the new one from L75
    stub_kind_get("Pod", items: [ready_pod_template]) # the new one from L76
    stub_kind_get("Node", items: three_node_templates) # maybe you can cheat and return the same one three times instead of having a separate fixture for this one too?
    ds.sync(build_resource_cache) # this puts the new data into the existing instance
```

Also, this is your regression test, so when you're done with it, make sure it fails if you revert your business logic change.

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 15]
```diff
@@ -38,22 +38,105 @@ def test_deploy_passes_when_updated_available_does_match
...
+  end
+
+  def test_deploy_fails_when_not_all_pods_ready
+    status = {
+      "desiredNumberScheduled": 3,
+      "updatedNumberScheduled": 3,
+      "numberReady": 2,
+    }
+    ds_template = build_ds_template(filename: 'daemon_set.yml', generation: 2, status: status)
+    pod_template = build_pod_template(filename: 'daemon_set_pods.yml')
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: yeah i did miss them originally, added them back now. I'll fix the verbiage to be more consistent with expectations.

**Comment by dalehamel** [comment]
@KnVerey when you have a chance I would appreciate another look :eyes:

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 16]
```diff
@@ -38,22 +38,99 @@ def test_deploy_passes_when_updated_available_does_match
...
+    ds = build_synced_ds(ds_template: ds_template, pod_templates: pod_templates, node_templates: node_templates)
+    refute_predicate(ds, :deploy_succeeded?)
+
+    node_added_status = {
+      "desiredNumberScheduled": 3,
+      "updatedNumberScheduled": 3,
+      "numberReady": 1,
+    }
+    ds_template = build_ds_template(filename: 'daemon_set.yml', status: node_added_status)
+    pod_templates = load_fixtures(filenames: ['daemon_set_pod.yml'])
```
Comment (KnVerey) [inline test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: For realism, shouldn't this return three pods, two of which are not ready? I think that right now, this test would still pass if the check were `@pods.all?(&:ready)`, unfiltered by node.

File: lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb [hunk 17]
```diff
@@ -38,6 +39,37 @@ def print_debug_logs?
...
+      def initialize(definition:)
+        @name = definition.dig("metadata", "name").to_s
+        @definition = definition
+      end
+    end
+
+    def relevant_pods_ready?
+      return true if rollout_data["desiredNumberScheduled"].to_i == rollout_data["numberReady"].to_i # all pods ready
+      relevant_node_names = @nodes.map(&:name)
+      considered_pods = @pods.select { |p| relevant_node_names.include?(p.node_name) }
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb]: i agree

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb (hunk 16 above)
Comment (dalehamel) [reply test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: I changed numberReady to 2, and made the fixture have 2 ready pods.

**Comment by dalehamel** [comment]
FYI @douglas as this might conflict with renaming in #585

**Comment by douglas** [comment]
Hey Dale! I think the other PR just needs reviews to go - if this is not urgent, can we ship it first? Then I volunteer to rebase this one.

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 18]
```diff
@@ -38,22 +38,99 @@ def test_deploy_passes_when_updated_available_does_match
...
+  end
+
   private
 
-  def build_ds_template(status: {})
-    base_ds_maifest = YAML.load_stream(File.read(File.join(fixture_path('for_unit_tests'), 'daemon_set.yml'))).first
-    base_ds_maifest.deep_merge("status" => status)
+  def build_ds_template(filename:, status: {})
+    base_ds_manifest = YAML.load_stream(File.read(File.join(fixture_path('for_unit_tests'), filename))).first
+    base_ds_manifest.deep_merge!("status" => status)
```
Comment (dturn) [inline test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: why the change to `deep_merge!`

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 19]
```diff
@@ -38,22 +38,99 @@ def test_deploy_passes_when_updated_available_does_match
...
+    ds_template = build_ds_template(filename: 'daemon_set.yml', status: node_added_status)
+    pod_templates = load_fixtures(filenames: ['daemon_set_pods.yml'])
+
+    stub_kind_get("DaemonSet", items: [ds_template])
+    stub_kind_get("Pod", items: pod_templates)
+    ds.sync(build_resource_cache)
+    assert_predicate(ds, :deploy_succeeded?)
+  end
+
+  def test_deploy_fails_when_not_ell_pods_updated
```
Comment (dturn) [inline test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: typo: `test_deploy_fails_when_not_ell_pods_updated` -> `test_deploy_fails_when_not_all_pods_updated`

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb [hunk 20]
```diff
@@ -38,22 +38,99 @@ def test_deploy_passes_when_updated_available_does_match
...
+      "updatedNumberScheduled": 1,
+    }
+    ds_template = build_ds_template(filename: 'daemon_set.yml', status: status)
+    pod_templates = load_fixtures(filenames: ['daemon_set_pod.yml'])
+    node_templates = load_fixtures(filenames: ['node.yml'])
+    ds = build_synced_ds(ds_template: ds_template, pod_templates: pod_templates, node_templates: node_templates)
+    assert_predicate(ds, :deploy_succeeded?)
+  end
+
+  def test_deploy_passes_when_ready_pods_but_node_added
```
Comment (dturn) [inline test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: `node_added` or `nodes_added` ?

File: test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb (hunk 18 above)
Comment (dalehamel) [reply test/unit/kubernetes-deploy/kubernetes_resource/daemon_set_test.rb]: artifact from debugging probably, i'll fix that.

File: lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb [hunk 21]
```diff
@@ -38,6 +39,37 @@ def print_debug_logs?
...
+      considered_pods.present? && considered_pods.all?(&:deploy_succeeded?)
+    end
+
+    def find_nodes(cache)
+      all_nodes = cache.get_all(Node.kind)
+
+      all_nodes.each_with_object([]) do |node_data, nodes|
```
[the remark this answers is not in the document]
Comment (dalehamel) [reply lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb]: yeah that's cleaner.

**Comment by dalehamel** [comment]
i double checked with @douglas and he said this is fine to merge, merging.
